### PR TITLE
Lemma 7.5.1

### PR DESCRIPTION
**Pull request by lakesare**
[[blueprint]](https://florisvandoorn.com/carleson/blueprint/treesection.html#dyadic-partition-1)

This can be merged!
PR review is appreciated.

### TODOs

- [x] prove the remaining `𝓘 u₁ ⊂ cube` subgoal (`lemma union_𝓙₅`)
- [x] prove the second theorem (`lemma pairwiseDisjoint_𝓙₅`)
- [x] add `\leanok`
- [x] refactor (`lemma union_𝓙₅`)
- [x] refactor (`lemma pairwiseDisjoint_𝓙₅`)
- [x] update the blueprint to not use strict containment 
- [x] address PR review suggestions

### CODE REVIEW

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 1]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+    · exact u₁_le.1
+    · exfalso
+      exact notDisjoint disjoint
+  · have weaker : s (𝓘 u₁) ≤ s cube := Int.le_of_lt (Int.lt_of_not_ge east)
+    apply GridStructure.fundamental_dyadic' at weaker
+    rw [disjoint_comm] at notDisjoint
+    exact Or.resolve_right weaker notDisjoint
+
+-- Auxiliary lemma for union_𝓙₅
+lemma exists_cube_in_𝓙_containing_point (hx: x ∈ (𝓘 u₁)) : ∃ cube ∈ 𝓙 (t.𝔖₀ u₁ u₂), x ∈ cube := by
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:67]: The canonical way to state this lemma is probably this: (not sure if that makes it easier or harder to use in your case)
```suggestion
lemma 𝓘_subset_sUnion_𝓙_𝔖₀ : 𝓘 u₁ ⊆ ⋃₀ (𝓙 (t.𝔖₀ u₁ u₂)) := by
```

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 2]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+  exact h
+
 /-- Part of Lemma 7.5.1. -/
 lemma union_𝓙₅ (hu₁ : u₁ ∈ t) (hu₂ : u₂ ∈ t) (hu : u₁ ≠ u₂)
     (h2u : 𝓘 u₁ ≤ 𝓘 u₂) :
     ⋃ J ∈ 𝓙₅ t u₁ u₂, (J : Set X) = 𝓘 u₁ := by
-  sorry
+  apply Set.Subset.antisymm
+  · rw [Set.subset_def]
+    intros x hx
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:80]:
```suggestion
  · intros x hx
```

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 3]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+      exact notDisjoint disjoint
+    · obtain ⟨p, belongs⟩ := t.nonempty' hu₁
+      by_contra! contr
+      have white := calc (𝓘 p : Set X)
+        _ ⊆ 𝓘 u₁ := if_descendant_then_subset t hu₁ belongs
+        _ ⊆ cube := by
+          apply u₁_is_subset_of_cube cube
+          · have notIn : cube ∉ t.𝓙₅ u₁ u₂ := λ a => contr cube a xInCube
+            unfold 𝓙₅ at notIn
+            rw [inter_def, Set.mem_setOf_eq, not_and_or] at notIn
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:113]:
```suggestion
            rw [𝓙₅, inter_def, Set.mem_setOf_eq, not_and_or] at notIn
```
I think this should work?

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 4]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+          unfold ball
+          rw [subset_def]
+          intro y xy
+          rw [mem_setOf_eq] at xy ⊢
+          have numbers : 4 * (D : ℝ) ^ s cube < 100 * D ^ (s cube + 1) := by
+            gcongr
+            linarith
+            exact one_lt_D (X := X)
+            linarith
+          exact gt_trans numbers xy
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:128]:
```suggestion
          exact numbers.trans xy
```

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 5]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+      have black : ¬↑(𝓘 p) ⊆ ball (c cube) (100 * ↑D ^ (s cube + 1)) := by
+        have in_𝔖₀ := 𝔗_subset_𝔖₀ (hu₁ := hu₁) (hu₂ := hu₂) (hu := hu) (h2u := h2u)
+        rw [subset_def] at in_𝔖₀
+        have tile := in_𝔖₀ p belongs
+        exact east p tile
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:133]:
```suggestion
        exact east p (in_𝔖₀ p belongs)
```
presumably this also works?

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 6]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
...
+lemma u₁_is_subset_of_cube
+  (cube : Grid X)
+  (h : cube ∉ Iic (𝓘 u₁))
+  (notDisjoint : ¬ Disjoint (cube : Set X) (𝓘 u₁ : Set X))
+  : (𝓘 u₁ : Set X) ⊆ cube := by
+  unfold Iic at h
+  rw [Set.nmem_setOf_iff, Grid.le_def, not_and_or] at h
+  cases' h with west east
+  · have h_le_cases := le_or_ge_or_disjoint (i:=cube) (j:=𝓘 u₁)
+    rcases h_le_cases with cube_le | u₁_le | disjoint
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:55]: You can also inline the `h_le_cases` here, i.e. `rcases le_or_ge_or_disjoint (i := cube) (j := 𝓘 u₁) ...`

File: Carleson/ForestOperator/LargeSeparation.lean [hunk 7]
```diff
@@ -42,17 +42,103 @@ def holderFunction (f₁ f₂ : X → ℂ)  (J : Grid X) (x : X) : ℂ :=
 
 /-! ### Subsection 7.5.1 and Lemma 7.5.2 -/
 
+-- Auxiliary lemma for union_𝓙₅
+lemma u₁_is_subset_of_cube
```
Comment (fpvandoorn) [inline Carleson/ForestOperator/LargeSeparation.lean:46]: We don't put variable names in lemma names. The way this lemma is formulated now can be something like `𝓘_subset_of_not_disjoint_of_nmem_Iic` or something.

You can generalize `𝓘 u₁` to an arbitrary element of `Grid X` and then move this lemma to `GridStucture`. 
Furthermore, I think you can proof this a bit more efficiently: don't do cases on `h` first, but do cases on `le_or_ge_or_disjoint` first.

**Review by fpvandoorn — COMMENTED** [review]
This looks good!
I mostly have some stylistic comments, 
I didn't open this PR in Lean myself to check whether all my suggestions work.

File: Carleson/ForestOperator/LargeSeparation.lean (hunk 7 above)
Comment (fpvandoorn) [reply Carleson/ForestOperator/LargeSeparation.lean:46]: About the name: that was just to explain the naming scheme; feel free to actually use a shorter name, (e.g. `Grid.subset_of_not_disjoint` after reformulation)

**Comment by lakesare** [comment]
Implemented the suggestions, they all worked.
Thanks for the review! This can be merged.

**Comment by fpvandoorn** [comment]
Thanks!

**Pull request merged**